code/notificationbuilder/src/main/java/com/adobe/marketing/mobile/notificationbuilder/internal/templates/AEPPushTemplate.kt

```diff
@@ -41,276 +40,184 @@ internal sealed class AEPPushTemplate {
-            intentExtras.getString(PushTemplateIntentConstants.IntentKeys.EXPANDED_BODY_TEXT)
-        expandedBodyTextColor =
-            intentExtras.getString(PushTemplateIntentConstants.IntentKeys.EXPANDED_BODY_TEXT_COLOR)
-        titleTextColor =
-            intentExtras.getString(PushTemplateIntentConstants.IntentKeys.TITLE_TEXT_COLOR)
-        smallIconColor =
-            intentExtras.getString(PushTemplateIntentConstants.IntentKeys.SMALL_ICON_COLOR)
-        notificationBackgroundColor =
-            intentExtras.getString(PushTemplateIntentConstants.IntentKeys.NOTIFICATION_BACKGROUND_COLOR)
-        isFromIntent = true
+    internal open fun createIntent(action: String): Intent {
+        val intent = Intent(action)
+        intent.flags = Intent.FLAG_ACTIVITY_SINGLE_TOP
+
+        // add the notification payload version
+        intent.putExtra(PushPayloadKeys.VERSION, payloadVersion?.toString())
+
+        // add the notification text information
+        intent.putExtra(PushPayloadKeys.TITLE, title)
+        intent.putExtra(PushPayloadKeys.BODY, body)
+        intent.putExtra(PushPayloadKeys.EXPANDED_BODY_TEXT, expandedBodyText)
+        intent.putExtra(PushPayloadKeys.TICKER, ticker)
+
+        // add the template type
+        intent.putExtra(PushPayloadKeys.TEMPLATE_TYPE, templateType?.value)
+
+        // add the basic media information
+        intent.putExtra(PushPayloadKeys.IMAGE_URL, imageUrl)
+
+        // add the notification action information
+        intent.putExtra(PushPayloadKeys.ACTION_URI, actionUri)
+        intent.putExtra(PushPayloadKeys.ACTION_TYPE, actionType?.name)
+
+        // add the notification icon information
+        intent.putExtra(PushPayloadKeys.SMALL_ICON, smallIcon)
+        intent.putExtra(PushPayloadKeys.LARGE_ICON, largeIcon)
+
+        // add the notification color data
+        intent.putExtra(PushPayloadKeys.TITLE_TEXT_COLOR, titleTextColor)
+        intent.putExtra(PushPayloadKeys.BODY_TEXT_COLOR, bodyTextColor)
+        intent.putExtra(PushPayloadKeys.BACKGROUND_COLOR, backgroundColor)
+        intent.putExtra(PushPayloadKeys.SMALL_ICON_COLOR, smallIconColor)
+
+        // add other notification properties
+        intent.putExtra(PushPayloadKeys.TAG, tag)
+        intent.putExtra(PushPayloadKeys.SOUND, sound)
+        intent.putExtra(PushPayloadKeys.BADGE_COUNT, badgeCount.toString())
+        intent.putExtra(PushPayloadKeys.STICKY, isNotificationSticky.toString())
+
+        // add notification priority and visibility
+        intent.putExtra(PushPayloadKeys.PRIORITY, priorityString)
+        intent.putExtra(PushPayloadKeys.VISIBILITY, visibilityString)
+
+        return intent
     }
 
-    fun getNotificationImportance(): Int {
-        notificationImportance?.let {
-            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) return notificationImportance as Int
-        }
-        return NotificationManager.IMPORTANCE_DEFAULT
+    fun getNotificationVisibility(): Int {
```

Comment:
Can you check why we need this? This function simply calls `NotificationVisibility.getNotificationCompatVisibilityFromString` without any additional logic

Reply:
If the common intent for notification action is created at a single place now, we shouldn't need this anymore

Reply:
After this change, I need to do a little more clean up on NotificationVisibility, NotificationImportance and NotificationPriority 

 We still use `getNotificationVisibility()` function still in 
LegacyNotificationBuilderClass
```
            .setSmallIcon(context, pushTemplate.smallIcon, pushTemplate.smallIconColor)
            // set notification visibility
            .setVisibility(pushTemplate.getNotificationVisibility())
            // add any action buttons defined for the notification
            .addActionButtons(
```

and in NotificationCompatBuilderExtension


To me, I think this method is still valid. So here is the workflow
AEPPushTempate reads the notification payload and stores

- priorityString : String
- visibilityString : String

these properties can be used directly when we build Intent values

and the method

getNotificationPriority() : Int
getNotificationVisibility() : Int
getNotificationImportance() : Int  Applicable for Android N and above

will give the ready to use values for the NotificationBuilders

Reply:
I will clean up a few more things around Visibility and Priority in my next PR. I will create a JIRA for it

Reply:
Ok but my comment was around why we need an extra function for it. The builder classes can directly call `NotificationVisibility.getNotificationCompatVisibilityFromString`. Or better, we can store the integer values for these in the template class and use `NotificationVisibility.getNotificationPriority` if/when we construct the intent